/*

leet code problem 701
    - Level - medium
    - Insert into a Binary Search Tree
    - https://leetcode.com/problems/insert-into-a-binary-search-tree/

You are given the root node of a binary search tree (BST) and a value to insert into the tree.
Return the root node of the BST after the insertion.
It is guaranteed that the new value does not exist in the original BST.

Notice that there may exist multiple valid ways for the insertion,
as long as the tree remains a BST after insertion. You can return any of them.


Example 1:
Input: root = [4,2,7,1,3], val = 5
Output: [4,2,7,1,3,5]

Example 2:
Input: root = [40,20,60,10,30,50,70], val = 25
Output: [40,20,60,10,30,50,70,null,null,25]

Example 3:
Input: root = [4,2,7,1,3,null,null,null,null,null,null], val = 5
Output: [4,2,7,1,3,5]

 */
package com.dsa.lc;

public class SearchBST {
    public TreeNode insertIntoBST(TreeNode root, int val) {
        if(root == null){
            root = new TreeNode(val);
            return root;
        }

        TreeNode temp = root;
        while(true){
            if(val < temp.val){
                if(temp.left != null){
                    temp = temp.left;
                }else{
                    temp.left = new TreeNode(val);
                    break;
                }
            }else{
                if(temp.right != null){
                    temp = temp.right;
                }else{
                    temp.right = new TreeNode(val);
                    break;
                }
            }
        }
        return root;
    }
}